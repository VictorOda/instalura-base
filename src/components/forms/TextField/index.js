import React from 'react';
import styled, { css } from 'styled-components';
import PropTypes from 'prop-types';
import get from 'lodash/get';
import Text from '../../foundation/Text';
import { propToStyle } from '../../../theme/utils/propToStyle';

const InputWrapper = styled.div`
  margin-bottom: 17px;
  ${propToStyle('width')}
  ${propToStyle('padding')}
`;

const Input = styled(Text)`
  width: 100%;
  border: 1px solid ${({ theme }) => get(theme, `colors.tertiaryLight.${theme.mode}.contrastText`)};
  border-radius: ${({ theme }) => theme.borderRadius};
  outline: none;

  ${({ theme, isFieldInvalid }) => isFieldInvalid && css`
    border-color: ${get(theme, `colors.error.${theme.mode}.color`)};
    & + span {
      color: ${get(theme, `colors.error.${theme.mode}.color`)};
      font-size: 11px;
    }
  `}

  ${propToStyle('padding')}
`;

Input.defaultProps = {
  tag: 'input',
  variant: 'paragraph1',
};

export default function TextField({
  placeholder,
  name,
  onChange,
  value,
  error,
  isTouched,
  padding,
  ...props
}) {
  const hasError = Boolean(error);
  const isFieldInvalid = hasError && isTouched;

  return (
    // eslint-disable-next-line react/jsx-props-no-spreading
    <InputWrapper {...props}>
      <Input
        type="text"
        placeholder={placeholder}
        name={name}
        onChange={onChange}
        value={value}
        isFieldInvalid={isFieldInvalid}
        padding={padding}
        // eslint-disable-next-line react/jsx-props-no-spreading
        {...props}
      />
      {isFieldInvalid
      && (
      <Text
        variant="smallestException"
        color="error"
        role="alert"
      >
        {error}
      </Text>
      )}
    </InputWrapper>
  );
}

TextField.defaultProps = {
  error: '',
  isTouched: false,
  padding: '12px 16px',
};

TextField.propTypes = {
  placeholder: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  value: PropTypes.string.isRequired,
  padding: PropTypes.string,
  error: PropTypes.string,
  isTouched: PropTypes.bool,
};
