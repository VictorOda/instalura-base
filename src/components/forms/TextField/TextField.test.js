import React from 'react';
import { render, screen } from '../../../infra/tests/testUtils';
import user from '@testing-library/user-event';
import TextField from './index';

describe('<TextField />', () => {
  test('renders component', () => {
    render(
      <TextField
        placeholder="Nome"
        value="Victor"
        onChange={() => {}}
        name="nome"
      />,
    );

    // screen.debug();

    const textField = screen.getByPlaceholderText(/nome/i);
    expect(textField).toMatchSnapshot();
  });

  describe('when field is valid', () => {
    describe('and user is typing', () => {
      test('the value must be updated', () => {
        const onChangeMock = jest.fn();
        render(
          <TextField
            placeholder="Nome"
            value="Victor"
            onChange={onChangeMock}
            name="nome"
            isTouched
          />,
        );

        const inputNome = screen.getByPlaceholderText(/nome/i);
        user.type(inputNome, 'teste@gmail.com');

        screen.debug();

        expect(onChangeMock).toHaveBeenCalledTimes(15);
      });
    });
  });

  describe('when field is invalid', () => {
    test('displays the respective error message', () => {
      render(
        <TextField
          placeholder="Email"
          value="teste@gmail.com"
          onChange={() => {}}
          name="email"
          isTouched
          error="O campo email é obrigatório"
        />,
      );

      const inputEmail = screen.getByPlaceholderText(/email/i);
      expect(screen.getByPlaceholderText(/email/i)).toHaveValue('teste@gmail.com');
      expect(screen.getByRole(/alert/i)).toHaveTextContent('O campo email é obrigatório');
      expect(inputEmail).toMatchSnapshot();
    });
  });
});
