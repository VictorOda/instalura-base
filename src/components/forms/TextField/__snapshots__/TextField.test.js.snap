// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> renders component 1`] = `
.c0 {
  font-size: 16px;
  font-weight: 400;
  line-height: 1.25;
  padding: 12px 16px;
}

.c1 {
  width: 100%;
  border: 1px solid #88989E;
  border-radius: 12px;
  outline: none;
  padding: 12px 16px;
}

<input
  class="c0 c1"
  name="nome"
  placeholder="Nome"
  type="text"
  value="Victor"
/>
`;

exports[`<TextField /> when field is invalid displays the respective error message 1`] = `
.c0 {
  font-size: 16px;
  font-weight: 400;
  line-height: 1.25;
  padding: 12px 16px;
}

.c1 {
  width: 100%;
  border: 1px solid #88989E;
  border-radius: 12px;
  outline: none;
  border-color: #dc3545;
  padding: 12px 16px;
}

.c1 + span {
  color: #dc3545;
  font-size: 11px;
}

<input
  class="c0 c1"
  name="email"
  placeholder="Email"
  type="text"
  value="teste@gmail.com"
/>
`;
